### ISSUE TITLE
Kubernetes - support local executors

### ISSUE THREAD

**Issue by florian-hoenicke** [jina-ai/jina#3223]
In the first Kubernetes iteration, it will just be possible to have containerized applications. Supported. Custom `python class Executors` are not supported in case they are not built to an image.

Please build them automatically.
The user should be able to specify the folder which contains
- python executor class
- additional python code
- requirements.txt

**Comment by numb3r3** [comment]
Why not ship this custom executor into our bubble system?

**Comment by JoanFM** [comment]
Isn't this already possible with hubble? does it belong in core?

**Comment by florian-hoenicke** [comment]
Okay, that should be sufficient for now.

**Comment by florian-hoenicke** [comment]
At some point it might be nice to even support local executors without creating them in hubble.

**Comment by 3cham** [comment]
@numb3r3, @JoanFM : I agree that shipping the executor through hubble would solve it, but it is not an easy job if you have several custom executors -> that usually involves maintaining one Dockerfile per executor. With every code change inside these executors, which happens quite often during the development process, we must build & push the images and deploy them. 

Testing the flow deployment would be slowed down pretty much. In contrast, if there is a properly defined base image, which could be built just once at the very beginning. The rest of the deployment process would only be:
- install new required libs
- copy updated source code

Another disadvantage might be, sometimes, user may not want to push their image if their source code contains proprietary work.

**Comment by tadejsv** [comment]
I think supporting local executors through kubernetes (and also JinaD) is something we definitely do not want, and should kill off altogether where it currently exist. If you support that, then you also have to deal with the headaches of transferring all the files over the network, supporting directory structures properly, installing python and system libraries, etc.

JinaD is already struggling with that, and in k8s these problems will probably explode. If someone is using k8s (or remote machines in particular), he should already be familiar with containers and container registries. So using custom executors can be easily achieved by creating docker containers and pushing them onto your registry (e.g. ECR). K8s already has the setup to pull containers from private registries, so this is really as simple as it can get.

**Comment by 3cham** [comment]
@tadejsv: thanks for sharing the insight. I originally thought JinaD has already found a good way to manage the dependencies & modules already. 

However, I still think the constraints that you mentioned could be solved by leveraging k8s API (cp, exec, etc...) in an **_acceptable_** way. This is really only for boosting the development process. For production deployment, I agree that there is no other way around shipping individual container images. 

Jina has delivered a great experience for local dev, so my initial thought was, it might be possible for deployment on k8s as well. Speaking of our own experience, about 70% of dev time we have been dealing with building & shipping containers. So we decided to use only one image across all the executors. which ends up having really large size. With a bit cache hacking, the image could be built & pushed quite fast, but that still feels uncomfortable though if we have to modify our executor a long the way.

**Comment by tadejsv** [comment]
@3cham why does building and shipping containers consume so much time in your case? What is the bottleneck there?

What JinaD is doing is reinventing the wheel of containerization - installing dependencies, copying over files - this is why conteinerization was invented. And not to mention that if you do this "local" development you will have to repeat the whole process before you go to production, because you need to make sure that what you developed locally (well, not really locally if you are using k8s), works in containers as well. If you are using containers during development (the k8s stage) as well, you won't have this problem

**Comment by JoanFM** [comment]
I think there is a missinterpretation on what the issue is trying to deal with here.

The issue claims that from the Jina K8s interface, one would provide simple executors and have jina be responsible of building the container. And the point would be if Jina offers via Hubble or does its own docker building and publish to registry for K8s to get them (Is that right @florian-hoenicke ?)

**Comment by 3cham** [comment]
@tadejsv: let's say you want to try different embedding methods, or modifying them with different hyper-parameters set, or maybe the way they are interacting with external storages / services and want to see the changes work on k8s. And just because we tried to use container from day 1, we had to build & push images extensively, for each modification. This is not the bottleneck but rather a repetitive task, that could be automated, but you still have to wait. 

The difficult part is not about having those code working locally but running on k8s performantly, because there, on k8s, is the whole infrastructure provided, and it is not really available for local development. Of course, this view is not based on the experience working with jina but on another toolings.

@JoanFM: no I didn't mean that jina must build the container. This idea is just about deploy a custom executor with previously built base image, and just by the deployment time, jina might overwrite / extend the entrypoint with installing dependencies, copying executor modules.
